### ISSUE TITLE
`.remove()` should allow removing all instances of a substring without using a RegExp

### ISSUE THREAD

**Issue by lolmaus** [andrewplummer/Sugar#473]
Hi!

It's frustrating that '.remove()` only removes the first instance.

The current workaround is to use a RegExp:

``` js
"foo".remove('o'); // => "fo"
"foo".remove(/o/g); // => "f"
```

I suggest adding a flag allowing it to remove all instances:

``` js
"foo".remove('o', true); // => "f"
```

Suggested implementation (might be buggy and/or slow):

``` js
  function(substr, all) {
    if (all && Object.isString(substr)) {
      substr = new RegExp(substr, 'g');
    }
    return this.replace(substr, '');
  }
```

**Comment by andrewplummer** [comment]
`remove` is really just a convenience method for `replace`, and if there's anywhere in the Sugar library whose value is questionable in the first place, this is probably it. I would be highly skeptical about the value of any deviation from acting like `replace`.

tl;dr Regexes are your friends, use them.

**Comment by lolmaus** [comment]
Isn't the whole point of Sugar to make developers' work easier and more enjoyable?

It's easier to write `arr.flatten()` than `[].concat.apply([], arr)` and that's why Sugar lets you use `flatten`.

It's easier to write `str.remove('o')` than `str.remove(/o/g)`, and that's why Sugar... doesn't let you use the easier way?

And i'm not asking to deviate from a standard. `remove` is a custom convenience method, and improving it like that will not violate any standard and is even unlikely to break anyone's code.

**Comment by andrewplummer** [comment]
> Isn't the whole point of Sugar to make developers' work easier and more enjoyable?

Not at the expense of causing confusion or unpredictable behavior. This is crucial and it's part of the reason that things can be enjoyable in the first place. This isn't an extreme example of that or anything but it's a balance that always needs to be kept in mind.

> It's easier to write str.remove('o') than str.remove(/o/g)

Is it? It's one more character.

> And i'm not asking to deviate from a standard.

No, but I could see someone being just as frustrated that it does global replacements when they just wanted to remove the first match.

**Comment by lolmaus** [comment]
> they just wanted to remove the first match

But why should `.remove` only remove first match by default? I believe that removing all matches is a much more common use case.

**Comment by andrewplummer** [comment]
It's a balance between common use case and expected behavior. Replace acts like this when strings are passed and remove is intended to be parallel to that. 

I'm also not convinced it's a "much more" common use case either.

**Comment by lolmaus** [comment]
Okay, i see your point, i'm sorry that i didn't see it earlier. I was thinking that `remove` has a standard counterpart and i didn't understand why a custom method should mimic an already existing one. But i finally realized that custom `remove` and standard `replace` do *different* things, i. e. `remove` supplements rather than replaces `replace`. And it would me much easier for everybody to use the same API on both of them.

But from the description of `remove` it is absolutely unclear that it will remove only the first occurrence, it's a big source of confusion for anyone who is not already familiar with `replace`. And if a person is already familiar with `replace`, he is likely to use `replace` rather than look up a convenience method on Sugar.

So my suggestions are:

* Explicitly document that `replace` only removes the first occurence of a string, while with a regex it removes either first or all occurence(s) based on the `g` flag. Provide examples.
* Create a `removeAll` method that removes all occurences of a given string. There are a number of concerns why this is important:
  * Having `remove` alongside `removeAll` will make it obvious that one of them does not remove all occurences of a string.
  * Not everyone is familiar with regexes and especially with their flags, and it will be much easier for those people to use `.removeAll(str)` than then figuring out regexes after stumbling over single replacement limitation.
  * You can't just put any string between slashes and have its regex counterpart. If your string contains many non-alphanumeric characters, you have to escape the shit out of it. That's tedious and absolutely unnecessary work.

**Comment by andrewplummer** [comment]
Yeah I mean I saw your point from the start. Kinda wish `replace` didn't work that way, and if I were to ever think to use a string with it I'd probably have an "oh yeah that's right" moment.

I will definitely update the documentation.

As far as `removeAll` goes I started out typing this response with the intention of being against it, but I'm starting to think about it more. I think I'll have to think more, but let me just response to each of the points.

> Having remove alongside removeAll will make it obvious...
Maybe... I guess. This one isn't so compelling to me tbh but you're not wrong.

> Not everyone is familiar with regexes and especially with their flags
This I completely disagree with. Here is my reason. While Sugar is intended to make things easier for developers I have been very conscious about not "masking" the correct way to do things. The problem with JS is that in many cases the "correct" way is often sloppy (creating edge cases) or long-winded and this is where Sugar can be of help. This is not the cases with regexes. They are unquestionably the correct way to do things in many cases. The methods are readable (`replace`, `match`) and make intent clear once you know what they do. They take a bit of effort to learn but that's only because they are so powerful. There are some things developers should never have to care how to do (like, say, create an array from an arguments object), but using regexes is not one of them.

> If your string contains many non-alphanumeric characters, you have to escape the shit out of it

This, however, is a much better argument, and very true. While, again, I'm not interested in creating shortcuts so that users don't have to learn regexes, you're still correct that escaping can be a pain. You've also got me thinking that a `replaceAll` method might also be very useful here where global replaces can be replaced with the arguments passed. `String#assign` exists but it can be clunky as you need to have constructed the string yourself ahead of time... So I'm thinking of something like this:

```
'The shirt costs $50 and the shoes $70'.replaceAll(/\$\d+/, '$80', '$100');
```

That would actually be rather difficult to pull off normally... in fact I'm kind of surprised this method hasn't made it in already...

Shit I started out against adding one method, now I'm thinking it might make sense to add two.......

**Comment by Mottie** [comment]
A `replaceAll` function wouldn't be too hard to add :wink: 

```js
// 'string'.replaceAll( /regex/, 'first replacement', 'second replacement', 'nth-replacement' );
String.prototype.replaceAll = function() {
	var indx = 0,
		replacements = Array.prototype.slice.call(arguments),
		regexp = replacements.shift();
	return this.replace( regexp, function(original) {
		// replace with original string if no replacement defined
		return replacements[indx++] || original;
	});
};
```

And using your example:

```js
'The shirt costs $50 and the shoes $70'.replaceAll(/\$\d+/g, '$80', '$100');
// "The shirt costs $80 and the shoes $100"
```

**Comment by andrewplummer** [comment]
Summarizing this issue:

--------------

`$50 and $100'.remove('$')` only removes the first instance of `$`.

@lolmaus claims that this is unexpected to those not familiar with regexes and also that doing it with regexes (`remove(/\$/g)`) is a pain to escape.

I disagree with the first point. Developers should familiarize themselves with regexes. As to the second point, this is not something that I see as being either all too common or such a large pain for smaller regexes, for example if it was just a dollar sign then it would be `.remove(/\$/g)`, which is hardly a big problem. However if this becomes a larger string then I begin to see his point, as it would be `.remove(new RegExp(RegExp.escape(largeStringWithTokens), 'g')`. This kind of slight code reduction with improved expressiveness is very much in line with Sugar's domain, however the use case is rather obscure. Very much on the fence about this one.

However, having a `replaceAll` method would also allow this syntax:
```
'$50 and $100'.replaceAll('$', '¥', '£');
```
This could quickly and easily allow multiple substitutes for each token matched, which is quite powerful. If `removeAll` was added as well it would allow parallelism that would 1) address @lolmaus problem with ambiguity and 2) handle complex escaping situations, which would address both his issues.

--------------

Looking at it again I think that overall there is enough here to make adding these methods beneficial, so I will add them. Improving the documents will also make the functionality even more clear.

**Comment by andrewplummer** [comment]
Ok, these two methods are now added! To see them all in a row:

- `String#replace` - Native JS method. Strings passed will be case-sensitive, escape tokens, and only replace one occurrence. Regexes will respect global flag.

- `String#remove` - Literally just calling `replace` with `''`.

- `String#replaceAll` - Replaces matches with arguments passed. Strings passed will be case-sensitive, escape tokens, and replace all occurrences. Regexes will have their global flag forced on.

- `String#removeAll` - Identical to `replaceAll` and will replace with `''`.

Documents updated to be more clear as well.

**Comment by lolmaus** [comment]
:+1:

**Comment by andrewplummer** [comment]
I'm going to leave this open for now until this actually goes out.
